### ISSUE TITLE
Only one networking interface being used

### ISSUE THREAD

**Issue by SubJunk** [UniversalMediaServer/UniversalMediaServer#687]
I use a USB wireless dongle as well as a wired network cable, and usually UMS sticks with whatever one gets used first. So if I find UMS with my phone, which is on the wireless network, UMS won't detect my TV which is on wired, and vice versa.
I haven't looked at the code so I'm not sure why it happens, but rarely it doesn't seem to happen, possible after time has passed.
Pretty poor information to start a bug report with, but it's a start and hopefully it can get more detailed.
I think solving this might help solve some of the connectivity issues that get reported.

**Comment by Nadahar** [comment]
This is probably different for HTTP v1 and v2 - which one do you use - and it would be nice if you tried the other one and compared the behaviour.

**Comment by SubJunk** [comment]
I was using v2 but I just confirmed it happens with both versions, no difference.

**Comment by Nadahar** [comment]
What would you consider to be the correct behavior in such a situation? You have two IP adresses on two different NIC's, would you want UMS to pick one of them or to actually respond to both? It could complicate things a bit, since it would mean that UMS would "own" multiple IP adresses.

**Comment by SharkHunter** [comment]
Looking at the code we always just pick one interface (for the multicast sockets). If you look at the code there is code to find/loop over all interfaces in the system but when multicastscokets should be created we only just pick one (like InetSocketAddress localAddress = new InetSocketAddress(usableAddresses.get(0), 0);
		MulticastSocket ssdpSocket = new MulticastSocket(localAddress);)

Now I can't really see the implications of just creating multicast sockets on all interfaces (and most I couldn't test it either since I only have one :)) but it should be ok just to have an array of interface sockets (or some sockopt or whatever haven't looked at the details of java multicastsocket).
The unicast sockets are most likely safe (or minimal effort to change) since they can be made listening on 0.0.0.0 instead of a interface address. 

Now all of the above should be addressed with some caution since binding to a particular interface (rather than all) is perhaps what some users want. It is a choice to have one interface on the DLNA network and one interface somewhere else.

**Comment by SubJunk** [comment]
We already have the ability in the user interface to force the use of one interface so I don't think we need to worry about that. It would be cool if we supported multiple by default as long as they can restrict it to just one, which they can. Our competitors use all interfaces I think (I tested and confirmed Serviio)

**Comment by SubJunk** [comment]
This is high on my wishlist, if you get time to try it @SharkHunter :)

**Comment by SubJunk** [comment]
@UniversalMediaServer/developers I've been trying to do this today and it seems like a lot of changes are needed to implement this, unless I don't understand

**Comment by Nadahar** [comment]
@SubJunk I think a broader approach is needed, if we're going to change anything about the network interface logic, we're going to have to fully understand the current logic first. I see shortcomings and bugs in the current logic, but I don't think this is a one day job.

**Comment by Nadahar** [comment]
@SubJunk Even though I don't know how or if it would work, I'm not convinced that uPnP/DLNA will handle operating on two subnets at once. The standard is very simplistic network wise, they basicly decided that "home users are so stupid we should only cater for the most basic scenario". That translates to just one subnet, and the whole standard is made with that in mind. Who in their right mind makes something in this millennium based on IP broadcast?

That doesn't mean we shouldn't try to look at the options, but the first thing I'm wondering is why you have different subnets for your wired and wireless network. Is it because you have different firewall rules/access from the two? There are so many "consumer" things that don't handle subnets properly that I gave that up a long time ago and put my wireless router in bridge/transparent/AP mode. If you do that, the whole problem goes away.

**Comment by SubJunk** [comment]
@Nadahar I haven't done any configuration of my network at all, I just have a USB Wi-Fi dongle plugged in via USB and my PS3 or TV (depending on which one I feel needs more testing at the time) plugged in via ethernet cable.

It seems that our code goes out of its way to avoid broadcasting on all interfaces at once, which is the default behavior for `InetSocketAddress()` and `MulticastSocket()`, so I think the approach that @SharkHunter suggested might work.

**Comment by Nadahar** [comment]
@SubJunk Are you saying that you don't have a wireless router? 

If you have a wireless router and had the router in bridged/transparent/AP mode you wouldn't need your dongle to reach your wireless devices, your wired and wireless network would be on the same subnet and every device, wired or not, would speak to eachother.

**Comment by SubJunk** [comment]
@Nadahar yeah I used a router like that at my old place but I'm keeping it like this until we have fixed this issue/added this feature. As long as it's annoying me it provides me motivation to fix it. We get reports on the forums of people having connection issues all the time and I'm happy to have reproduced one of the more common ones now.

Some people don't have routers and we know that other media servers support this setup so we should, too.

**Comment by Nadahar** [comment]
@SubJunk Agree, as I said I'm not saying we shouldn't fix it, I'm just a bit scared of "rush fixes" to core things like network.

**Comment by SubJunk** [comment]
@Nadahar yeah I agree, we should tread carefully. Also hopefully @skeptical will pop in soon since he has spent a fair amount of time in that area of the code so I hope he has some insights that can help us.

**Comment by Nadahar** [comment]
@SubJunk I'm in the final stages of pushing the final freedesktop.org branch (Linux trash support) and making a PR.

Once that's done, I'll have to finish the "external dependency" red tape and get the artifact published on central, so that we can integrate it and get travis up and working.

When that's done I should go back to the crowdin branch and finish that, but I've seen another thing that creates "connection issues" on more than one occation now - and I'm thinking about looking at that before finishing the crowdin branch. I've been involved in more than one case where the run-one-instance function doesn't work, and particularly the fact that it doesn't check if UMS is running as a service. I think a lot of those in this situation has at some point pressed the "install as a service" button without having any idea what that means, and from that moment on they will always have a headless UMS running in the "background" that start with Windows. Whenever they start UMS from the desktop after that, they will have a lot of issues. 

I'd like to finish up some of the things, I hate having things lying around getting out of sync, but at the same time I think this is a somewhat urgent issue to tackle.

**Comment by SubJunk** [comment]
We have a new forum thread where a user is being affected by this feature gap: http://www.universalmediaserver.com/forum/viewtopic.php?f=9&t=5809

**Comment by colemickens** [comment]
How does one workaround this? I was hoping for the usual flag to specify the listen interface, but it doesn't seem like it exists. Is this low-hanging fruit that would be an easy PR for me to do?

**Comment by Nadahar** [comment]
I don't know if it's that low-hanging. Getting unicast to use all network interfaces should be as simple as using 0.0.0.0. The problem, AFAICR, is multicast. upon which UPnP relies. If my memory serves me correct, we'd have to run one "server instance" per network interface that we want to handle multicast on. We'd then have to organize how to handle the incoming packages from the different listening servers, and then to respond on the correct interface. It would take some logistics to get this to work.
